plugins/bkpr/descriptions.c

```diff
@@ -0,0 +1,323 @@
+{
+	return &phd->payment_hash;
+}
+
+static bool payment_hash_key_eq(const struct payment_hash_desc *phd,
+				const struct sha256 *h)
+{
+	return sha256_eq(&phd->payment_hash, h);
+}
+
+HTABLE_DEFINE_NODUPS_TYPE(struct payment_hash_desc,
+			  payment_hash_key,
+			  hash_payment_hash,
+			  payment_hash_key_eq,
+			  payment_hash_desc_htable);
+
+
+/* We keep descriptions in memory and datastore */
+struct descriptions {
+	struct utxo_desc_htable *by_utxo;
+	struct payment_hash_desc_htable *by_payment_hash;
+};
+
+static const char *ds_desc_utxo_path(const tal_t *ctx,
+				     const struct bitcoin_outpoint *outp)
+{
+	return tal_fmt(ctx, "bookkeeper/description/utxo/%s", fmt_bitcoin_outpoint(tmpctx, outp));
+}
+
+static void utxo_desc_datastore_update(struct command *cmd,
+				       const struct utxo_desc *utxo_desc)
+{
+	const char *path = ds_desc_utxo_path(tmpctx, &utxo_desc->outp);
+
+	jsonrpc_set_datastore_binary(cmd, path,
+				     utxo_desc->desc, strlen(utxo_desc->desc),
+				     "create-or-replace",
+				     ignore_datastore_reply, NULL, NULL);
+}
+
+static const char *ds_desc_payment_hash_path(const tal_t *ctx,
+						    const struct sha256 *payment_hash)
+{
+	return tal_fmt(ctx, "bookkeeper/description/payment/%s",
+		       fmt_sha256(tmpctx, payment_hash));
+}
+
+static void payment_hash_desc_datastore_update(struct command *cmd,
+					       const struct payment_hash_desc *phd)
+{
+	const char *path = ds_desc_payment_hash_path(tmpctx, &phd->payment_hash);
+	jsonrpc_set_datastore_binary(cmd, path, phd->desc, strlen(phd->desc),
+				     "create-or-replace",
+				     ignore_datastore_reply, NULL, NULL);
+}
+
+static struct payment_hash_desc *
+new_payment_hash_description(struct descriptions *descriptions,
+			     const struct sha256 *payment_hash,
+			     const char *desc TAKES)
+{
+	struct payment_hash_desc *phd;
+
+	phd = payment_hash_desc_htable_get(descriptions->by_payment_hash, payment_hash);
+	if (phd) {
+		tal_free(phd->desc);
+		phd->desc = tal_strdup(phd, desc);
+	} else {
+		phd = tal(descriptions->by_payment_hash, struct payment_hash_desc);
+		phd->payment_hash = *payment_hash;
+		phd->desc = tal_strdup(phd, desc);
+		payment_hash_desc_htable_add(descriptions->by_payment_hash, phd);
+	}
+	return phd;
+}
+
+void add_payment_hash_description(struct command *cmd,
+				  struct bkpr *bkpr,
+				  const struct sha256 *payment_hash,
+				  const char *desc TAKES)
+{
+	struct payment_hash_desc *phd;
+
+	phd = new_payment_hash_description(bkpr->descriptions,
+					   payment_hash, desc);
+	payment_hash_desc_datastore_update(cmd, phd);
+}
+
+static struct utxo_desc *
+new_utxo_description(struct descriptions *descriptions,
+		     const struct bitcoin_outpoint *outpoint,
+		     const char *desc TAKES)
+{
+	struct utxo_desc *ud;
+
+	ud = utxo_desc_htable_get(descriptions->by_utxo, outpoint);
+	if (ud) {
+		tal_free(ud->desc);
+		ud->desc = tal_strdup(ud, desc);
```

Comment:
we say we're `TAKES` the `desc` in the method params, but we strdup it here (and below)

Reply:
tal_strdup() takes :)